VIM-7487: Adds support for connected apps

jasonhawkins opened the pull request:
### Ticket
[VIM-7487](https://vimean.atlassian.net/browse/VIM-7487)

### Pull Request Checklist
- [x] Resolved any merge conflicts
- [x] No build errors or warnings are introduced
- [x] New files are written in Swift
- [x] New classes contain license headers
- [x] New classes have Documentation
- [x] New public methods have Documentation

### Ticket Summary
Update VimeoNetworking to include support for the new Connected Apps endpoints.

### Implementation Summary

#### ConnectedApp
An object that contains all of the data related to a connection to a specific platform. However, not all of the properties relate to all platforms. Exceptions are noted in the documentation where applicable. Generally, `ConnectedApp` includes data related to when the connection was established, the platform it represents, and the scopes needed to interact with a particular feature. For now features are limited to live streaming and publishing to social.

#### Request+ConnectedApp
Adds convenience methods for constructing requests related to connected apps. This includes requesting all existing connected apps, creating a connection to a specific platform, and deleting an existing connection.

#### VIMConnection
Adds `VIMConnectionNameConnectedApps`, allowing a client to access the connected apps connection.

#### VIMInteraction
Expanded to support VIMUser interactions related to connected apps. You can now look for interactions on a user for each of the supported platforms (`facebook_connected_app`, for instance) to get information about whether the app is already connected, and what scopes are required for each of the supported features.

### Reviewer Tips
Testing this work requires access to the internal branch, `https://connected-apps-endpoints-api.ci.vimeows.com`. Furthermore, these features will only available to whitelisted apps for the foreseeable future once the changes are merged into the API.

### How to Test
Ensure unit tests pass. If testing against the CI branch ensure your client application is able to make the included requests and parse the results.

jasonhawkins committed:
Add connected apps to VIMInteraction

jasonhawkins committed:
Add connected apps to VIMConnection

jasonhawkins committed:
Add ConnectedApp model

jasonhawkins committed:
Add ConnectedApp request extension

jasonhawkins committed:
Add ConnectedAppTests

jasonhawkins committed:
Add ConnectedApp request extension tests

antonklevets commented on Sources/Shared/Request+ConnectedApp.swift:
```diff
@@ -0,0 +1,81 @@
+        var tokenKey: String
+        switch appType {
+        case .facebook:
+            tokenKey = .accessToken
```
https://github.vimeows.com/Vimeo-API/design/pull/146/files#diff-a9ed519c3fbb374ef99e26b80498642fR25
For facebook should use `.authCode`

antonklevets replied:
PR Fix https://github.com/vimeo/VimeoNetworking/pull/392

jasonhawkins replied:
Thanks, @antonklevets. I've updated this branch so that Facebook uses `auth_code` and updated the related test.

jasonhawkins committed:
Update Facebook token key to use auth_code

jasonhawkins committed:
Update connected app Facebook request test

jasonhawkins committed:
Update documentation

cameo-github commented:
<!--
  0 Errors
  1 Warning: Skipping duplicate build file ...
  1 Message: Executed 211 tests, with 0 fai...
  1 Markdown
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Skipping duplicate build file in Compile Sources build phase: /Users/distiller/VimeoNetworking/Sources/Shared/Requests/Request+Apple.swift (in target ‘VimeoNetworking-iOS’ from project ‘VimeoNetworking’)</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 211 tests, with 0 failures (0 unexpected) in 1.011 (1.334) seconds</td>
    </tr>
  </tbody>
</table>

See build details on [CircleCI](https://circleci.com/gh/vimeo/VimeoNetworking/1619)

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

jasonhawkins committed:
Fix typo

jasonhawkins committed:
Merge branch 'develop' into vim-7487-add-connected-apps

jasonhawkins commented:
build this please

jasonhawkins committed:
Add additional json fixtures

This will allow us to test the connect_apps and individual connected app responses

jasonhawkins committed:
Resolve issues related to object mapping

jasonhawkins committed:
Add additional tests covering object mapping

jasonhawkins committed:
Merge branch 'develop' into vim-7487-add-connected-apps

jasonhawkins committed:
Fix identifier type for PublishOptionItem

jasonhawkins committed:
Update and expand ConnectedApp parsing tests

jasonhawkins committed:
Add test page to facebook response fixture

jasonhawkins committed:
Change pages type from [String] to [PublishOptionItem]

jasonhawkins committed:
Add pages tests for facebook connected app

jasonhawkins committed:
Fix typo in documentation

Commit:
Merge branch 'develop' of https://github.com/vimeo/VimeoNetworking into vim-7487-add-connected-apps

* 'develop' of https://github.com/vimeo/VimeoNetworking:
  Turn `isActive` into readonly property (#400)
  Move specific Video file interfaces into public headers for MacOS & tvOS targets.
  Add Video files interfaces into framework umbrella header.
  Add isActive member on VIMPictureCollection (#398)
  Move specific Video file interfaces into public headers.
  Add method for creating a request that will create a thumbnail on a video (#396)

Commit:
Merge branch 'vim-7487-add-connected-apps' of https://github.com/vimeo/VimeoNetworking into vim-7487-add-connected-apps

* 'vim-7487-add-connected-apps' of https://github.com/vimeo/VimeoNetworking:
  Fix typo in documentation
  Add pages tests for facebook connected app
  Change pages type from [String] to [PublishOptionItem]
  Add test page to facebook response fixture
  Update and expand ConnectedApp parsing tests
  Fix identifier type for PublishOptionItem
  Add additional tests covering object mapping
  Resolve issues related to object mapping
  Add additional json fixtures

jasonhawkins committed:
Merge branch 'develop' into vim-7487-add-connected-apps

jasonhawkins committed:
Merge branch 'develop' into vim-7487-add-connected-apps

jasonhawkins committed:
Resolve project.pbxproj conflicts

rpassis commented on Sources/Shared/Models/ConnectedApp.swift:
```diff
@@ -0,0 +1,237 @@
+    case twitter
+    case youtube
+
+    public var stringValue: String {
```
how about conforming this to `CustomStringConvertible` and changing the naming here to `description` so it conforms? I'm guessing it would need to be done in an extension to avoid the @objc annotation but that should be fine since this property is not exposed to Objective-C runtime.

rpassis commented on Sources/Shared/Models/ConnectedApp.swift:
```diff
@@ -0,0 +1,237 @@
+public class ConnectedAppScopes: VIMModelObject {
+
+    /// All scopes required for publishing to a specific social media platform.
+    public var publishToSocial: [String]?
```
Is there a finite list of scopes that we can enumerate to make this a bit more type safe? At the moment I'm not even sure how to format this `var` with the different scopes - is it comma separate, spaces etc?

rpassis replied:
Also does it make sense for all of these properties to be publicly writeable as well? I'm not sure how they are going to be used but the less mutability of these models the better.

jasonhawkins replied:
The actual scopes aren't defined in the client app, instead they come from the server. We get them by looking for an interaction on the user object for a platform, `facebook_connected_app`, for example. The returned `VIMInteraction` will contain the scopes needed for the supported "connected app" features.

This class is supposed to encapsulate and store those scopes on a `VIMInteraction` object. 

```ObjC
@interface VIMInteraction : VIMModelObject
…
@property (nonatomic, strong, nullable) ConnectedAppScopes *allScopes;`
```

But I think you're right about the access for these properties though. I'll take a look.

rpassis commented on Sources/Shared/Models/ConnectedApp.swift:
```diff
@@ -0,0 +1,237 @@
+    public var publishToSocial: [String]?
+
+    /// All scopes required for simulcasting to a specific social media platform.
+    public var simulcast: [String]?
```
same as above

rpassis commented on Sources/Shared/Models/ConnectedApp.swift:
```diff
@@ -0,0 +1,237 @@
+public class PublishOptionItem: VIMModelObject {
+
+    /// The ID of the publish item.
+    public var identifier: String?
```
Same re: public setters - I'm guessing the identifier should not be publicly writeable?

rpassis commented on Sources/Shared/Models/ConnectedApp.swift:
```diff
@@ -0,0 +1,237 @@
+    /// The ID of the publish item.
+    public var identifier: String?
+
+    /// The name or display name of the publich item, i.e.: "art", "family", "vacation" etc.
```
typo on `publish`

rpassis commented on Sources/Shared/Models/ConnectedApp.swift:
```diff
@@ -0,0 +1,237 @@
+    /// The date when the user established this connected app, as a String.
+    /// - Note: This property is available to provide interoperability with Objective-C codebases.
+    ///         Using `addDate` is preferred.
+    @objc public var addDateString: String?
```
Same comment re: public setters

rpassis commented on Sources/Shared/Models/ConnectedApp.swift:
```diff
@@ -0,0 +1,237 @@
+    /// Facebook only. A value of 1 maps to expired data access. A value of 0 maps to having data access.
+    /// - Note: This property is available to provide interoperability with Objective-C codebases.
+    ///         Using `isDataAccessExpired` is preferred.
+    @objc public var dataAccessIsExpired: NSNumber?
```
does this need to be public given the computed boolean below?

jasonhawkins replied:
Good call, it does not.

rpassis commented on Sources/Shared/Models/VIMInteraction.h:
```diff
@@ -53,21 +59,33 @@ typedef NS_ENUM(NSInteger, VIMInteractionStreamStatus) {
+ For connected app interactions this indicates whether the app is connected. A value of 0 or `nil` means no
+ connection has been established, or an existing one has expired.
+ */
+@property (nonatomic, strong, nullable) NSNumber *isConnected;
```
could this be made private and exposed as a real (computed) boolean?

rpassis commented on Sources/Shared/Requests/Request+ConnectedApp.swift:
```diff
@@ -0,0 +1,76 @@
+//  THE SOFTWARE.
+//
+
+public extension Request {
```
should this be constrained to the `ConnectApp` type with `where ModelType ==`? I'm not sure how the type inference would work here without it - I'm guessing the consumer of the API would have to explicitly specify the request type otherwise?

rpassis commented on Sources/Shared/Requests/Request+ConnectedApp.swift:
```diff
@@ -0,0 +1,76 @@
+    /// - Parameters:
+    ///   - appType: The app platform for which the connection will be established.
+    ///   - token: An authentication token from the provided platfrom, used to establish the connection.
+    static func connect(to appType: ConnectedAppType, with token: String) -> Request {
```
Is it worth making this request more generic? As in do we always expect a token to be the method of authenticating with a third party? Could there be cases where we need to pass more than 1 parameter?

jasonhawkins replied:
For the platforms we support we only require the token for authentication. I guess it's possible that some future platform might have different requirements? But I think for now this should be fine.

rpassis reviewed (changes requested):
This looks good Jason, I love how thoroughly tested it is! 

I've left a few comments, mostly regarding visibility & availability of certain properties for you to consider. Let me know what you think!

jasonhawkins committed:
Add resourceKey to ConnectedApp object

jasonhawkins committed:
Update JSON fixtures with resource_key

jasonhawkins committed:
Update ConnectedApp tests to validate resourceKey

jasonhawkins committed:
Update ConnectedAppType to conform to CustomStringConvertible

jasonhawkins committed:
Update ConnectedApp requests

jasonhawkins committed:
Update tests to use description instead of stringValue

jasonhawkins committed:
Update access control for Connected App model properties

jasonhawkins committed:
Fix typo in documentation

jasonhawkins committed:
Change dataAccessIsExpired access to private, update tests

jasonhawkins committed:
Replace isConnected property with instance method

jasonhawkins committed:
Add tests for isConnectedToApp method

jasonhawkins committed:
Constrain request type to ModelType: ConnectedApp

jasonhawkins commented:
Thanks for the feedback @rpassis, this is ready for another look 👀

antonklevets commented on Sources/Shared/Requests/Request+ConnectedApp.swift:
```diff
@@ -23,7 +23,7 @@
 //
 
-public extension Request {
+public extension Request where ModelType: ConnectedApp {
```
@jasonhawkins In this case model for `connectedApps()` method is wrong and always empty. There should be array of `ConnectedApp`

I used this code
```
let request = Request<[ConnectedApp]>.connectedApps()
....
case .success(let response):
   response.model // Array of ConnectedApp
```
and was all fine.

antonklevets replied:
Also for `deleteConnectedApp` seems should be `VIMNullResponse` model instead of `ConnectedApp`, because there no info in response, i.e. i always receive error with message `VimeoClient requestSuccess returned invalid/absent dictionary`

should be 
`Request<VIMNullResponse>.deleteConnectedApp(.facebook)`

jasonhawkins replied:
Hey Anton, you're totally right. Sorry about that. I've just pushed changes that will constrain the request objects to the correct types, and added some tests to validate those types.

You should now be able to use those request objects without having to specify `<[ConnectedApp]>` or `<VIMNullResponse>`, etc.

jasonhawkins committed:
Constrain ConnectedApp requests to correct types

jasonhawkins committed:
Clean up @objc and @nonobjc annotation

jasonhawkins committed:
Merge branch 'develop' into vim-7487-add-connected-apps

rpassis commented on Sources/Shared/Models/ConnectedApp.swift:
```diff
@@ -0,0 +1,244 @@
+    /// The type of the connected app.
+    /// - Note: A returned value of `nil` indicates an unsupported `ConnectedAppType`, or a malformed response.
+    @nonobjc public var type: ConnectedAppType? {
+        switch self.typeString {
```
What do you think of moving this to the `ConnectedAppType` instead? Does it feel weird to conform to RawRepresentable `String` when the backing type of the enum is `Int`?

```
extension ConnectedAppType: RawRepresentable {
    public typealias RawValue = String
    public var rawValue: RawValue { return String(describing: self) }
    public init?(rawValue: Self.RawValue) {
        switch rawValue { ... }
    }
}
```

jasonhawkins replied:
That's a good suggestion, thanks. I've change the `type var` implementation to this:

```Swift
@nonobjc public var type: ConnectedAppType? {
    typeString.map { ConnectedAppType.init(rawValue: $0) } ?? nil
}
```

So we can still return nil if something goes wrong, or we encounter an unsupported platform. 

But the backing type for `ConnectedAppType` is an Int only for Objective-C interoperability.

rpassis replied:
Could probably go point free `typestring.map(ConnectedAppType.init)`

jasonhawkins replied:
I'd still need to `?? nil` otherwise I'm trying to return an optional optional.

rpassis commented on Sources/Shared/Models/VIMInteraction.m:
```diff
@@ -81,14 +93,18 @@ - (void)setStreamStatus {
 }
 
+- (BOOL)isConnectedToApp {
+    return self.isConnected == TRUE;
```
replace `TRUE` with `YES`

rpassis reviewed (approved):
@jasonhawkins Two comments for you to consider, and feel free to merge once they are addressed!

jasonhawkins committed:
Replace use of TRUE with YES

jasonhawkins committed:
Improve rawValue handling for ConnectedAppType
